### DISCUSSION TITLE
Best practices for larger freestanding projects

### DISCUSSION THREAD

**Discussion by real-tintin** [zephyrproject-rtos/zephyr#55642] [Q&A]
Hi!

I'm wondering what the best practices for larger (more than a few SW components) freestanding projects are. I have read the [Application Development](https://docs.zephyrproject.org/3.2.0/develop/application/index.html) which presents the different structures e.g., freestanding project with a src/main and how they are build using Cmake.

However, now assume a larger project with some freestanding SW components e.g.,

```
larger-freestanding-application/
├─── component_one/
    ├── inc/
    ├── src/
    ├── tests/
    └── CMakeLists.txt
├─── component_two/
    ├── inc/
    ├── src/
    ├── tests/
    └── CMakeLists.txt
 .
 .
 .
├─── component_n/
├─── main.c
└─── CMakeLists.txt
```

This is a structure which I have commonly encountered and used in previous project (*but not using Cmake - so bear with me).

So, is this this also feasible to use for a freestanding Zephyr application? In my mind each feature is its own library. Is that possible, assuming that they depend on the zephyr kernel itself? I found this sample application in the repo _samples/application_development/external_lib/CMakeLists.txt_. But it is independent of the zephyr kernel.

I haven't look to much at the ZTest framework yet, more than played about with the samples. But it would be nice if they could be structured as above i.e., each unit test tests each component isolated.

Cheers,

Oscar

**Comment by DerekSnell** [comment]
Hi @real-tintin ,

Yes, having the individual components like that is feasible.  I spend some time working on free-standing apps outside of the tree.  I have always linked the whole app as a single library, like the Zephyr samples.  But you could link each component as a library.

I am far from being an expert on CMake.  I just search the CMake files in the Zephyr tree to find examples, and then hack my file until it does what I want.  But the CMake files in the tree do provide many good references.  And if you build a Zephyr sample and look at the generated .MAP file, it shows lots of component libraries built from the tree, and then linked to the sample app.  

I do use Kconfigs in my app to include the different components.  This makes it easier to develop or debug one component at a time, and then integrate them together when ready.  You can add a file `Kconfig` in the larger-freestanding-application to define your Kconfig symbols.  And then select them in your prj.conf file.  Then your primary CMakeLists.txt can conditionally include the components selected.

I have little experience with ZTest or Twister, so I will let someone else comment about that.  

Best regards

**Reply by real-tintin** [reply]
Hi @DerekSnell,

Thanks for the quick feedback! That sound promising then. I have not used the Kconfig framework (yet). Sounds like a good companion.

After some reading and YouTubing about CMake I managed to put together a (almost) working example for the above use case. BUT...without support to make use of the <zephyr/kernel.h> in a library... When building it throws the following:

`_#include <zephyr/kernel.h> Doesn't work atm to build lib with zephyr dependencies_`

So well, I guess back to the drawing board.

For future reference. Below are the two minimal CMakeLists.txt I mad so far, and they work if one doesn't have any dependencies to the Zephyr kernel.

CMakeLists.txt in **top** dir:
```
cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(simple_app)

add_subdirectory(component_one_lib)
add_subdirectory(component_two_lib)

target_sources(app PRIVATE
    main.c
)

target_link_libraries(app PUBLIC component_one_lib)
target_link_libraries(app PUBLIC component_two_lib)
```

CMakeLists.txt in a **component** dir:
```
cmake_minimum_required(VERSION 3.20.0)

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(freestanding_app)

add_library(component_one STATIC
    src/component_one.c
)

target_include_directories(component_one_lib PUBLIC inc)
```

**Comment by pdgendt** [comment]
Hi @real-tintin, you might be interested in the "snippets" proposal, see #53640

**Comment by benediktibk** [comment]
[utils.zip](https://github.com/zephyrproject-rtos/zephyr/files/10982944/utils.zip)

This is a minimal example for a utils-library. You can then include it in your main application in CMakeLists.txt:

	cmake_minimum_required(VERSION 3.21.1)

	set(ZEPHYR_EXTRA_MODULES
		$ENV{ZEPHYR_BASE}/../libs/utils
	)

	find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
	project(awesome_stuff)

	message(STATUS "define sources for main app")

	target_sources(app PRIVATE
		src/main.c
	)

And activate it in your prj.conf:

        CONFIG_UTILS=y

**Reply by real-tintin** [reply]
Thanks for the example @benediktibk, much appriciated! I managed to build your utils example.

However, I tried to setup a unit-test using ZTest, but stumbled upon some issues. I expanded your structure to this:

```
├── CMakeLists.txt
├── main.c
├── prj.conf
└── utils
    ├── CMakeLists.txt
    ├── Kconfig
    ├── include
    │   └── blub.h
    ├── src
    │   └── blub.c
    ├── tests
    │   ├── CMakeLists.txt
    │   ├── prj.conf
    │   ├── test_blub.c
    │   └── testcase.yaml
    └── zephyr
        └── module.yml
```

In utils/tests/CMakeLists.txt I tried to mimic the CMakeLists.txt in the root, but when I run:

`west twister --platform unit_testing -T ./utils_project/`

I get this:

`utils_project/utils/tests/prj.conf:1: warning: attempt to assign the value 'y' to the undefined symbol UTILS`

Should it be possible to use the ZTest framework to test a zephyr lib? I'm not that familiar with KConfig.

Here is the whole project: [utils_project.zip](https://github.com/zephyrproject-rtos/zephyr/files/10988861/utils_project.zip)

**Reply by benediktibk** [reply]
Are you certain that $ENV{STANDALONE_ROOT}/utils_project/utils expands in the context of twister to the correct path? For testing purposes, a test project can also be build like a normal application using `west build`. This might give you some clues.

**Reply by real-tintin** [reply]
> Are you certain that $ENV{STANDALONE_ROOT}/utils_project/utils expands in the context of twister to the correct path? For testing purposes, a test project can also be build like a normal application using `west build`. This might give you some clues.

Sorry for my late reply. 

I believe I have, I manage to build the application from the CMakeLists.txt in the project root, but not when building the test application. I have not dugg deeper into why.

My current build chain only builds object files, I have attached a minimalistic example which builds and runs unit tests as well.
[simple_project.zip](https://github.com/zephyrproject-rtos/zephyr/files/11070924/simple_project.zip)

Note, I'm not happy with how includes are handle in some of the unit tests, as they are relative. I think I will restructure it to use absolute path's from the project root. More explicit.

**Comment by real-tintin** [comment]
I have now re-arranged the structure somewhat to promote explicit "absolute" (actually relative to top dir) include paths. I think this is also make is overall more readable to maintain source and include files along each other, and instead split on component level:

```
├── calculator
│   ├── operations
│   │   ├── tests
│   │   │   ├── CMakeLists.txt
│   │   │   ├── prj.conf
│   │   │   ├── test_add.c
│   │   │   ├── test_subtract.c
│   │   │   └── testcase.yaml
│   │   ├── CMakeLists.txt
│   │   ├── add.c
│   │   ├── add.h
│   │   ├── subtract.c
│   │   └── subtract.h
│   ├── tests
│   │   ├── CMakeLists.txt
│   │   ├── prj.conf
│   │   ├── test_calculator.c
│   │   └── testcase.yaml
│   ├── CMakeLists.txt
│   ├── calculator.c
│   └── calculator.h
├── logger
│   ├── integration_tests
│   │   ├── CMakeLists.txt
│   │   ├── prj.conf
│   │   ├── test_logger.c
│   │   └── testcase.yaml
│   ├── tests
│   │   ├── CMakeLists.txt
│   │   ├── prj.conf
│   │   ├── test_logger.c
│   │   └── testcase.yaml
│   ├── CMakeLists.txt
│   ├── logger.c
│   └── logger.h
├── CMakeLists.txt
├── README.md
├── main.c
├── prj.conf
└── sample.yaml
```
